File: edition4/js/ch01/1-1.js
function hasAllUnique(str) {
  var hasSeen = new Array(256);

  for (var i=0; i<str.length; i+=1) {
    var code = str.charCodeAt(i);
    if (hasSeen[code]) {
      return false;
    }

    hasSeen[code] = true;
  }

  return true;
}

// tests
var assert = require('assert');

describe('hasAllUnique', function() {

  it('should fail', function() {
    assert.equal(hasAllUnique('aa'), false);
    assert.equal(hasAllUnique('abcdefghijklmnopqrstuvwxyza'), false);
  });

  it('should succeed', function() {
    assert.equal(hasAllUnique('abcdefghijklmnopqrstuvwxyz'), true);
  });

});

File: edition4/c/ch01/1-1.cpp
#include <assert.h>

#include <algorithm>
#include <bitset>
#include <cmath>
#include <iostream>
#include <string>

using namespace std;

bool hasAllUnique(string str) {
  bitset<245> seen;

  for (int i=0; i<str.length(); i++) {
    if (seen.test(str[i]) == true) {
      return false;
    }

    seen.set(str[i], true);
  }

  return true;
}

int main() {
  string str1 = string("a");
  assert(hasAllUnique(str1));

  string str2 = string("abcdefghijklmnopqrstuvwxyz");
  assert(hasAllUnique(str2));

  string str3 = string("aa");
  assert(!hasAllUnique(str3));

  string str4 = string("abcdefghijklmnopqrstuvwxyza");
  assert(!hasAllUnique(str4));

  return 0;
}
